src/libaktualizr/uptane/uptanerepository.cc

```diff
@@ -117,113 +100,93 @@ bool Repository::verifyMetaTargets(const Uptane::Targets& director_targets, cons
-  new_meta.image_root = Uptane::Root(now, "repo", fetchRole(config.uptane.repo_server, Role::Root()), meta_.image_root);
-
-  new_meta.image_timestamp = Uptane::TimestampMeta(
-      now, "repo", fetchRole(config.uptane.repo_server, Role::Timestamp(), Version()), meta_.image_root);
-  if (new_meta.image_timestamp.version() > meta_.image_timestamp.version()) {
-    new_meta.image_snapshot = Uptane::Snapshot(
-        now, "repo", fetchRole(config.uptane.repo_server, Role::Snapshot(), Version()), meta_.image_root);
-    new_meta.image_targets = Uptane::Targets(
-        now, "repo", fetchRole(config.uptane.repo_server, Role::Targets(), Version()), meta_.image_root);
-  } else {
-    new_meta.image_snapshot = meta_.image_snapshot;
-    new_meta.image_targets = meta_.image_targets;
+  // should be already put to the storage by fetcher
+  if (!storage->loadUncheckedMetadata(&meta_raw)) {
+    LOG_WARNING << "No metadata to check";
+    return false;
   }
 
-  new_meta.director_targets = Uptane::Targets(
-      now, "director", fetchRole(config.uptane.director_server, Role::Targets(), Version()), meta_.director_root);
-  if (new_meta.director_root.version() > meta_.director_root.version() ||
-      new_meta.image_root.version() > meta_.image_root.version() ||
-      new_meta.director_targets.version() > meta_.director_targets.version() ||
-      new_meta.image_timestamp.version() > meta_.image_timestamp.version()) {
-    if (verifyMetaTargets(new_meta.director_targets, new_meta.image_targets)) {
-      storage->storeMetadata(new_meta);
-      meta_ = new_meta;
-    } else {
-      LOG_WARNING << "Metadata image/directory repo consistency check failed.";
+  try {
+    // first check with previous root
+    new_meta.director_root =
+        Uptane::Root(now, "director", Utils::parseJSON(meta_raw.director_root), meta_.director_root);
+    new_meta.image_root = Uptane::Root(now, "repo", Utils::parseJSON(meta_raw.image_root), meta_.image_root);
+
+    // second check against the new root
+    new_meta.director_root =
+        Uptane::Root(now, "director", Utils::parseJSON(meta_raw.director_root), new_meta.director_root);
+    new_meta.image_root = Uptane::Root(now, "repo", Utils::parseJSON(meta_raw.image_root), new_meta.image_root);
+
+    new_meta.image_timestamp =
+        Uptane::TimestampMeta(now, "repo", Utils::parseJSON(meta_raw.image_timestamp), new_meta.image_root);
+    new_meta.image_snapshot =
+        Uptane::Snapshot(now, "repo", Utils::parseJSON(meta_raw.image_snapshot), new_meta.image_root);
+    new_meta.image_targets =
+        Uptane::Targets(now, "repo", Utils::parseJSON(meta_raw.image_targets), new_meta.image_root);
+
+    new_meta.director_targets =
+        Uptane::Targets(now, "director", Utils::parseJSON(meta_raw.director_targets), new_meta.director_root);
+
+    if (new_meta.director_root.version() < meta_.director_root.version() ||
```

Comment:
It is legal to load a root but not a timestamp. Also, when a root is rotated, the protocol say to blow away all stored metadata, so the root **MUST** always go up, but the others **MAY** go down if and only if the root was rotated.

Reply:
> the others MAY go down if and only if the root was rotated.

Is it UPTANE-compatible?

Reply:
Yes, that is in the TUF spec.